src/dataflow/src/render/reduce.rs

```diff
@@ -144,94 +143,111 @@ where
                             output.push((key.clone(), 1));
                         }
                     }),
                     err_input,
-                )
-            } else if aggregates.len() == 1 {
-                // If we have a single aggregate, we need not stage aggregations separately.
-                (
-                    build_aggregate_stage(ok_input, 0, &aggregates[0], true, *monotonic),
-                    err_input,
-                )
+                );
+                let index = (0..group_key.len()).collect::<Vec<_>>();
+                self.set_local_columns(relation_expr, &index[..], (oks, errs.arrange()));
             } else {
-                // We'll accumulate partial aggregates here, where each contains updates
-                // of the form `(key, (index, value))`. This is eventually concatenated,
-                // and fed into a final reduce to put the elements in order.
-                let mut ok_partials = Vec::with_capacity(aggregates.len());
-                // Bound the complex dataflow in a region, for better interpretability.
-                scope.region(|region| {
-                    // Create an iterator over collections, where each is the application
-                    // of one aggregation function whose results are annotated with its
-                    // position in the final results. To be followed by a merge reduction.
-                    for (index, aggr) in aggregates.iter().enumerate() {
-                        // Collect the now-aggregated partial result, annotated with its position.
-                        let ok_partial = build_aggregate_stage(
-                            ok_input.enter(region),
+                // Collect aggregates with their indexes, so they can be sliced and diced.
+                let mut accumulable = Vec::new();
+                let mut remaining = Vec::new();
+                for index in 0..aggregates.len() {
+                    if accumulable_hierarchical(&aggregates[index].func).0 {
+                        accumulable.push((index, aggregates[index].clone()));
+                    } else {
+                        remaining.push((index, aggregates[index].clone()));
+                    }
+                }
+
+                let arrangement = if remaining.is_empty() {
+                    // If we have only accumulable aggregations, they can be arranged and returned.
+                    build_accumulables(ok_input, accumulable, true)
+                } else if remaining.len() == 1 && accumulable.is_empty() {
+                    // If we have a single non-accumulable aggregation, it can be arranged and returned.
+                    build_aggregate_stage(ok_input, 0, &aggregates[0], true, *monotonic)
+                } else {
+                    // Otherwise we need to stitch things together.
+                    let mut to_collect = Vec::new();
+                    let accumulable = build_accumulables(ok_input.clone(), accumulable, false)
```

Comment:
it seems like we should only build this dataflow if accumulable is non-empty? My understanding is that the tradeoff here is marginally more code complexity + an extra branch at render time, but the cost is an extra dataflow fragment thats never used